### PR TITLE
JIT: ARM64 SVE format encodings, `SVE_HY_3A` to `SVE_IB_3A`

### PR DESCRIPTION
**Pull request by TIHan**
Contributes to https://github.com/dotnet/runtime/issues/94549

Adds formats:
- SVE_HY_3A
- SVE_HY_3A_A
- SVE_HY_3B
- SVE_IB_3A
- SVE_HZ_2A_B
- SVE_IA_2A

Left: Capstone,
Right: Jit
image

### CODE REVIEW

File: src/coreclr/jit/emit.h [hunk 1]
```diff
@@ -1007,6 +1007,16 @@ class emitter
                 regNumber      _idReg4 : REGNUM_BITS;
             };
 
+            struct
```
Comment (TIHan) [inline src/coreclr/jit/emit.h:1010]: @dotnet/jit-contrib I'm not proud of doing this.

The idea is to use the space that stores `_idReg4` but use it for `insSvePrfop`.

I tried doing a union inside the struct above like so:
```cpp
            struct
            {
                // This 32-bit structure can pack with these unsigned bit fields
                emitLclVarAddr iiaLclVar;
                unsigned       _idRegBit : 1; // Reg3 is scaled by idOpSize bits
                GCtype         _idGCref2 : 2;
                regNumber      _idReg3 : REGNUM_BITS;
                union
                {
                    regNumber _idReg4 : REGNUM_BITS;
                    insSvePrfop _idSvePrfop : REGNUM_BITS;
                }
            };
```

But this fails a compile-time assert:
`    C_ASSERT(sizeof(emitter::instrDesc) == (SMALL_IDSC_SIZE + sizeof(void*)));`

This tells me that the size of the `instrDesc` changed when I introduced the nested union. Perhaps it is due to alignment? Is there an alternative way?

**Comment by TIHan** [comment]
@dotnet/jit-contrib @dotnet/arm64-contrib @kunalspathak @a74nh this is ready.

File: src/coreclr/jit/emit.h (hunk 1 above)
Comment (BruceForstall) [reply src/coreclr/jit/emit.h:1010]: IIUC, C++ is very picky about packing bitfields in that they all need to be of the same type. The "union" probably breaks that. You could have the `idSvePrfop` functions just use `_idReg4` directly, maybe with casting. Still ugly :-)

Comment (EgorBo) [reply src/coreclr/jit/emit.h:1010]: From my understanding `instrDesc` must be 8 bytes. We sometimes allocate more space for it for Debug or if JitDisasm is enabled in order to store more info for every record +8 byte -- I guess you can go similar route or make a side dictionary? Presumably, unconditional extension for instrDesc from 8 bytes to 16 will be a major TP regression. Afair, it was already painful to increase bit width for `_idIns` field because of SVE

File: src/coreclr/jit/codegenarm64test.cpp [hunk 2]
```diff
@@ -7511,6 +7511,118 @@ void CodeGen::genArm64EmitterUnitTestsSve()
                               INS_SCALABLE_OPTS_UNPREDICATED);
     theEmitter->emitIns_R_R_I(INS_sve_str, EA_SCALABLE, REG_V2, REG_R3, 255, INS_OPTS_NONE,
                               INS_SCALABLE_OPTS_UNPREDICATED);
+
+    // IF_SVE_HY_3A
+    theEmitter->emitIns_PRFOP_R_R_R(INS_sve_prfb, EA_SCALABLE, SVE_PRFOP_PLDL1KEEP, REG_P1, REG_R2, REG_V3,
```
Comment (a74nh) [inline src/coreclr/jit/codegenarm64test.cpp:7516]: When I was doing `ptrue`, that uses an enum. I considered squashing the enum into `INS_SCALABLE_OPTS`. But, here we require explicit sopt values. Therefore it's best to keep as is

tldr: all good :)

File: src/coreclr/jit/emit.h (hunk 1 above)
Comment (a74nh) [reply src/coreclr/jit/emit.h:1010]: I prefer the casting option over the version in the PR. You could even go as far as renaming `_idReg4` and making it unsigned instead of `regNumber` so that it's explicit it's used for multiple things.

If it's alignment causing the issues, then have you tried this way:
```
            struct
            {
                emitLclVarAddr iiaLclVar;
                union
                {
                    regNumber _idReg4 : REGNUM_BITS;
                    insSvePrfop _idSvePrfop : REGNUM_BITS;
                };
                unsigned       _idRegBit : 1; // Reg3 is scaled by idOpSize bits
                GCtype         _idGCref2 : 2;
                regNumber      _idReg3 : REGNUM_BITS;

            };
```
`emitLclVarAddr` is 32bits, so this will put  `_idReg4/_idSvePrfop` to 32bit alignment. Or if that doesn't work, try pushing the union to the top of the struct.


`// This 32-bit structure can pack with these unsigned bit fields` - This comment doesn't make sense as `emitLclVarAddr` is 32bits, making the full struct 47bits. Probably best to remove. It's left over from when the Arm32 and Arm64 code was entangled.


Whatever solution we end up with, in a later PR we should refactor `_idSvePattern` so that it uses the same mechanism.

File: src/coreclr/jit/emitarm64.cpp [hunk 3]
```diff
@@ -18792,52 +19110,89 @@ void emitter::emitIns_Call(EmitCallType          callType,
 
 /*****************************************************************************
  *
- *  Returns the encoding for the immediate value as 5-bits at bit locations '20-16'.
+ *  // Returns the encoding for the immediate value that is a multiple of 2 as 5-bits at bit locations '20-16'.
```
Comment (a74nh) [inline src/coreclr/jit/emitarm64.cpp:19113]: Did you mean to add the `//` ?

File: src/coreclr/jit/emitarm64.cpp [hunk 4]
```diff
@@ -29338,6 +29938,165 @@ emitter::insExecutionCharacteristics emitter::getInsExecutionCharacteristics(ins
             result.insLatency    = PERFSCORE_LATENCY_2C;
             break;
 
+        case IF_SVE_HY_3A: // .........h.mmmmm ...gggnnnnn.oooo -- SVE 32-bit gather prefetch (scalar plus 32-bit scaled
```
Comment (a74nh) [inline src/coreclr/jit/emitarm64.cpp:29941]: These are all 1C/1C and can be squashed into a single case

File: src/coreclr/jit/codegenarm64test.cpp [hunk 5]
```diff
@@ -7511,6 +7511,118 @@ void CodeGen::genArm64EmitterUnitTestsSve()
...
+                                    INS_OPTS_SCALABLE_S_SXTW); // PRFB    <prfop>, <Pg>, [<Xn|SP>, <Zm>.S, <mod>]
+    theEmitter->emitIns_PRFOP_R_R_R(INS_sve_prfb, EA_SCALABLE, SVE_PRFOP_CONST7, REG_P1, REG_R2, REG_V3,
+                                    INS_OPTS_SCALABLE_S_SXTW); // PRFB    <prfop>, <Pg>, [<Xn|SP>, <Zm>.S, <mod>]
+    theEmitter->emitIns_PRFOP_R_R_R(INS_sve_prfb, EA_SCALABLE, SVE_PRFOP_CONST14, REG_P1, REG_R2, REG_V3,
+                                    INS_OPTS_SCALABLE_S_UXTW); // PRFB    <prfop>, <Pg>, [<Xn|SP>, <Zm>.S, <mod>]
+    theEmitter->emitIns_PRFOP_R_R_R(INS_sve_prfb, EA_SCALABLE, SVE_PRFOP_CONST15, REG_P1, REG_R2, REG_V3,
+                                    INS_OPTS_SCALABLE_S_UXTW); // PRFB    <prfop>, <Pg>, [<Xn|SP>, <Zm>.S, <mod>]
+
+    // IF_SVE_HY_3A_A
+    theEmitter->emitIns_PRFOP_R_R_R(INS_sve_prfb, EA_SCALABLE, SVE_PRFOP_PLDL1KEEP, REG_P1, REG_R2, REG_V3,
```
Comment (a74nh) [inline src/coreclr/jit/codegenarm64test.cpp:7559]: nit: everything from here downwards uses `SVE_PRFOP_PLDL1KEEP`

Comment (a74nh) [inline src/coreclr/jit/codegenarm64test.cpp:7559]: nit: everything from here downwards uses SVE_PRFOP_PLDL1KEEP

File: src/coreclr/jit/emitarm64.cpp (hunk 4 above)
Comment (amanasifkhalid) [reply src/coreclr/jit/emitarm64.cpp:29941]: We have quite a few SVE cases that can be squashed. Should we enforce cleaning these up proactively as we add encodings, or should we expect to do this cleanup later? I'm fine either way.

File: src/coreclr/jit/emitarm64.h [hunk 6]
```diff
@@ -701,6 +715,24 @@ static bool isValidSimm4_MultipleOf32(ssize_t value)
     return (-256 <= value) && (value <= 224) && (value % 32 == 0);
 };
 
+// Returns true if 'value' is a legal signed multiple of 2 immediate 5 bit encoding (such as for LD1H).
```
Comment (amanasifkhalid) [inline src/coreclr/jit/emitarm64.h:718]: Nit: these comments should probably say "unsigned" instead of "signed".

File: src/coreclr/jit/instrsarm64sve.h [hunk 7]
```diff
@@ -285,31 +285,31 @@ INST6(prfb,              "prfb",                  0,                       IF_SV
     // PRFB    <prfop>, <Pg>, [<Xn|SP>, <Zm>.S, <mod>]                                   SVE_HY_3A           100001000h1mmmmm 000gggnnnnn0oooo     8420 0000   
     // PRFB    <prfop>, <Pg>, [<Xn|SP>, <Zm>.D, <mod>]                                   SVE_HY_3A_A         110001000h1mmmmm 000gggnnnnn0oooo     C420 0000   
     // PRFB    <prfop>, <Pg>, [<Xn|SP>, <Zm>.D]                                          SVE_HY_3B           11000100011mmmmm 100gggnnnnn0oooo     C460 8000   
-    // PRFB    <prfop>, <Pg>, [<Zn>.D{, #<imm>}]                                         SVE_HZ_2A_B         10000100000iiiii 111gggnnnnn0oooo     8400 E000   
+    // PRFB    <prfop>, <Pg>, [<Zn>.S{, #<imm>}]                                         SVE_HZ_2A_B         10000100000iiiii 111gggnnnnn0oooo     8400 E000
```
Comment (amanasifkhalid) [inline src/coreclr/jit/instrsarm64sve.h:288]: Nice catch

File: src/coreclr/jit/emitarm64.cpp [hunk 8]
```diff
@@ -14475,6 +14522,205 @@ void emitter::emitIns_R_PATTERN_I(instruction ins, emitAttr attr, regNumber reg1
...
+        case INS_sve_prfw:
+        case INS_sve_prfd:
+            assert(isLowPredicateRegister(reg1));
+            assert(isScalableVectorSize(size));
+
+            if (isVectorRegister(reg2))
+            {
+                assert(insOptsScalableWords(opt));
+
+#ifdef DEBUG
```
Comment (amanasifkhalid) [inline src/coreclr/jit/emitarm64.cpp:14668]: Not a nit, but I've seen both ways, so I'm curious: For control flow like this that just asserts something, what is the consensus on `ifdef`'ing it, versus expecting the compiler to remove it as dead code in Release builds?

**Review by amanasifkhalid — COMMENTED** [review]
LGTM, will approve once the `_idSvePrfop/_idReg4` refactor is in.

File: src/coreclr/jit/emit.h (hunk 1 above)
Comment (TIHan) [reply src/coreclr/jit/emit.h:1010]: Yea, I'm definitely leaning towards doing that, it's not super great to have a bunch of dummy padding fields.

File: src/coreclr/jit/emitarm64.h (hunk 6 above)
Comment (TIHan) [reply src/coreclr/jit/emitarm64.h:718]: Yea it should say unsigned.

File: src/coreclr/jit/emitarm64.cpp (hunk 8 above)
Comment (TIHan) [reply src/coreclr/jit/emitarm64.cpp:14668]: It's very possible that it would all be removed without the `ifdef`. But putting the `ifdef` around it helps me know the difference between code that is used to verify or not.

File: src/coreclr/jit/codegenarm64test.cpp (hunk 2 above)
Comment (TIHan) [reply src/coreclr/jit/codegenarm64test.cpp:7516]: Yea, I considered `INS_SCALABLE_OPTS`, but it would explode in the number of cases since some of these instructions need to take UXTX and SXTX.

File: src/coreclr/jit/emitarm64.cpp (hunk 4 above)
Comment (TIHan) [reply src/coreclr/jit/emitarm64.cpp:29941]: Since these values are not right and need to be updated anyway, I tend not to squash them as the instructions could be different when we go back and fix these up.

File: src/coreclr/jit/emitarm64.cpp (hunk 3 above)
Comment (TIHan) [reply src/coreclr/jit/emitarm64.cpp:19113]: I didn't, will remove them.

**Comment by TIHan** [comment]
@amanasifkhalid this is ready again

**Review by amanasifkhalid — APPROVED** [review]
LGTM, thanks!

**Review by a74nh — APPROVED** [review]
LGTM
